Failed tests (11121): TestStoreRangeDownReplicate

The following test appears to have failed:

[#11121](https://circleci.com/gh/cockroachdb/cockroach/11121):

```
I0112 12:58:42.596862 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I0112 12:58:42.597042 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I0112 12:58:42.597192 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I0112 12:58:42.597337 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I0112 12:58:42.597506 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestStoreRangeDownReplicate (10.44s)
	client_raft_test.go:693: Failed to achieve proper replication within 10 seconds
=== RUN   TestChangeReplicasDescriptorInvariant
E0112 12:58:42.601827 926 gossip/gossip.go:210  different node IDs were set for the same gossip instance (2147483647, 1)
I0112 12:58:42.602395 926 raft/raft.go:446  [group 1] 1 became follower at term 5
I0112 12:58:42.602669 926 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0112 12:58:42.602809 926 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0112 12:58:42.602933 926 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0112 12:58:42.603066 926 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0112 12:58:42.603210 926 raft/raft.go:487  [group 1] 1 became leader at term 6
I0112 12:58:42.604701 926 storage/replica.go:1604  gossiping cluster id  from store 1, range 1
--
I0112 12:58:57.848837 926 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestLeaderAfterSplit (0.25s)
	<autogenerated>:28: prior leak detected, leaktest disabled
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.81s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	33.342s
=== RUN   TestBatchBasics
I0112 12:58:52.859422 989 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchGet
I0112 12:58:52.866982 989 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I0112 12:58:52.875317 989 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.01s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -550,7 +550,7 @@ func TestStoreRangeUpReplicate(t *testing.T) {
 	wg.Wait()
 
 	// Once we know our peers, trigger a scan.
-	mtc.stores[0].ForceReplicationScan()
+	mtc.stores[0].ForceReplicationScanAndProcess()
 
 	// The range should become available on every node.
 	if err := util.IsTrueWithin(func() bool {
@@ -629,62 +629,6 @@ func TestStoreRangeDownReplicate(t *testing.T) {
 	}
 	wg.Wait()
 
-	// storeIDset is used to compare the replica sets from different views (i.e.
-	// local range descriptors)
-	type storeIDset map[roachpb.StoreID]struct{}
-	makeStoreIDset := func(replicas []roachpb.ReplicaDescriptor) storeIDset {
-		idSet := make(storeIDset)
-		for _, r := range replicas {
-			idSet[r.StoreID] = struct{}{}
-		}
-		return idSet
-	}
-
-	// Function to see if the replication level of the new range has reached the
-	// expected equilibrium. If equilibrium has not been reached, this function
-	// returns the list of stores that *should* have a replica for the range.
-	checkReplication := func() (bool, storeIDset) {
-		// Query each store for a replica of the range, generating a real map of
-		// the replicas.
-		foundIDset := make(storeIDset)
-		foundLocalRangeDescs := make([]*roachpb.RangeDescriptor, 0, len(mtc.stores))
-		for _, s := range mtc.stores {
-			r := s.LookupReplica(keys.Addr(splitKey), nil)
-			if r != nil {
-				foundLocalRangeDescs = append(foundLocalRangeDescs, r.Desc())
-				foundIDset[s.StoreID()] = struct{}{}
-			}
-		}
-
-		// Fail immediately if there are less than three replicas.
-		replicaCount := len(foundIDset)
-		if replicaCount < 3 {
-			t.Fatalf("Removed too many replicas; expected at least three replicas, found %d", replicaCount)
-		}
-
-		// Look up the official range descriptor, make sure it agrees with the
-		// found replicas.
-		realRangeDesc := getRangeMetadata(keys.Addr(rightKey), mtc, t)
-		realIDset := makeStoreIDset(realRangeDesc.Replicas)
-		if !reflect.DeepEqual(realIDset, foundIDset) {
-			return false, realIDset
-		}
-
-		// Ensure the local range descriptors everywhere agree with reality.
-		for _, desc := range foundLocalRangeDescs {
-			localIDset := makeStoreIDset(desc.Replicas)
-			if !reflect.DeepEqual(localIDset, foundIDset) {
-				return false, realIDset
-			}
-		}
-
-		// If we have only three replicas, exit the loop.
-		if replicaCount == 3 {
-			return true, nil
-		}
-		return false, foundIDset
-	}
-
 	maxTimeout := time.After(10 * time.Second)
 	succeeded := false
 	for !succeeded {
@@ -693,24 +637,20 @@ func TestStoreRangeDownReplicate(t *testing.T) {
 			t.Fatalf("Failed to achieve proper replication within 10 seconds")
 		case <-time.After(10 * time.Millisecond):
 			mtc.expireLeaderLeases()
-			// If our replication level matches the target, we have succeeded.
-			var idSet storeIDset
-			succeeded, idSet = checkReplication()
-			if succeeded {
+			rangeDesc := getRangeMetadata(keys.Addr(rightKey), mtc, t)
+			if count := len(rangeDesc.Replicas); count < 3 {
+				t.Fatalf("Removed too many replicas; expected at least 3 replicas, found %d", count)
+			} else if count == 3 {
+				succeeded = true
 				break
 			}
 
-			// Kick off a manual ReplicaGC Scan on any store which is not part of the
-			// current replica set. Kick off a Replication scan on *one* store which
-			// is part of the replica set.
-			kickedOffReplicationQueue := false
+			// Run replication scans on every store; only the store with the
+			// leader lease will actually do anything. If we did not wait
+			// for the scan to complete here it could be interrupted by the
+			// next call to expireLeaderLeases.
 			for _, store := range mtc.stores {
-				if _, ok := idSet[store.StoreID()]; !ok {
-					store.ForceReplicaGCScanAndProcess()
-				} else if !kickedOffReplicationQueue {
-					store.ForceReplicationScan()
-					kickedOffReplicationQueue = true
-				}
+				store.ForceReplicationScanAndProcess()
 			}
 		}
 	}
@@ -1208,8 +1148,8 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 				mtc.stores[1].GossipStore()
 			})
 			// Force the repair queues on all alive stores to run.
-			mtc.stores[0].ForceReplicationScan()
-			mtc.stores[1].ForceReplicationScan()
+			mtc.stores[0].ForceReplicationScanAndProcess()
+			mtc.stores[1].ForceReplicationScanAndProcess()
 		}
 	}
 	ticker.Stop()
@@ -1283,7 +1223,7 @@ func TestStoreRangeRebalance(t *testing.T) {
 			}
 
 			mtc.expireLeaderLeases()
-			mtc.stores[1].ForceReplicationScan()
+			mtc.stores[1].ForceReplicationScanAndProcess()
 		}
 	}
 }

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -56,7 +56,7 @@ import (
 	"github.com/gogo/protobuf/proto"
 )
 
-const replicationTimeout = 3 * time.Second
+const replicationTimeout = 5 * time.Second
 
 // Check that Stores implements the RangeDescriptorDB interface.
 var _ kv.RangeDescriptorDB = &storage.Stores{}

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -276,7 +276,20 @@ func (m *multiTestContext) Stop() {
 		m.reenableTableSplits()
 	}()
 
-	<-done
+	select {
+	case <-done:
+	case <-time.After(30 * time.Second):
+		// If we've already failed, just attach another failure to the
+		// test, since a timeout during shutdown after a failure is
+		// probably not interesting, and will prevent the display of any
+		// pending t.Error. If we're timing out but the test was otherwise
+		// a success, panic so we see stack traces from other goroutines.
+		if m.t.Failed() {
+			m.t.Error("timed out during shutdown")
+		} else {
+			panic("timed out during shutdown")
+		}
+	}
 }
 
 // rpcSend implements the client.rpcSender interface. This implementation of "rpcSend" is

--- a/storage/helpers_test.go
+++ b/storage/helpers_test.go
@@ -21,14 +21,16 @@
 
 package storage
 
-// ForceReplicationScan iterates over all ranges and enqueues any that
-// need to be replicated. Exposed only for testing.
-func (s *Store) ForceReplicationScan() {
+// ForceReplicationScanAndProcess iterates over all ranges and
+// enqueues any that need to be replicated. Exposed only for testing.
+func (s *Store) ForceReplicationScanAndProcess() {
 	s.mu.Lock()
-	defer s.mu.Unlock()
 	for _, r := range s.mu.replicas {
 		s.replicateQueue.MaybeAdd(r, s.ctx.Clock.Now())
 	}
+	s.mu.Unlock()
+
+	s.replicateQueue.DrainQueue(s.ctx.Clock)
 }
 
 // DisableReplicaGCQueue disables or enables the replica GC queue.

--- a/storage/store.go
+++ b/storage/store.go
@@ -1307,9 +1307,11 @@ func (s *Store) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.Bat
 				resolveIntents, pErr = s.resolveWriteIntentError(ctx, wiErr, rng, args, ba.Header, pushType)
 				if len(resolveIntents) > 0 {
 					if resErr := rng.resolveIntents(ctx, resolveIntents, false /* !wait */, true /* poison */); resErr != nil {
-						// When resolving asynchronously, should never get an error
-						// back here.
-						panic(resErr)
+						// When resolving asynchronously, errors should not
+						// usually be returned here, although there are some cases
+						// when they may be (especially when a test cluster is in
+						// the process of shutting down).
+						log.Warningf("asynchronous resolveIntents failed: %s", resErr)
 					}
 				}
 				// Make sure that if an index is carried in the error, it

